Admin client returns HTTP code `400 Bad Request` when using x509 certificate

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/client-java

### Describe the bug

Good day folks!

As topic says, whenever I create a client with x509 certificate credentials, I get HTTP 400 both on client and Keycloak.

I have created the simplest setup environment to reproduce. See [here](https://github.com/igpetrov/kc-x509).

Consider the following client created:

```
final Keycloak keycloakClient = KeycloakBuilder.builder()
        .serverUrl("https://localhost:18443/auth")
        .realm("master")
        .grantType(OAuth2Constants.CLIENT_CREDENTIALS)
        .clientId("qwerty")
        .clientSecret("xxxxxxx")
        .resteasyClient(resteasyClient)
        .build();
    
    // dummy call
    final var serverInfo = keycloakClient.serverInfo().getInfo();
```

On the Keycloak:

```
2024-10-10 13:00:19 2024-10-10 10:00:19,754 DEBUG [org.hibernate.internal.util.EntityPrinter] (executor-thread-35) org.keycloak.models.jpa.entities.ClientAttributeEntity{value=CN=Private, OU=Private, O=Private, L=Odesa, ST=ODS, C=UA, EMAILADDRESS=igpetrov@github.io}
2024-10-10 13:00:19 2024-10-10 10:00:19,754 DEBUG [org.hibernate.orm.sql.exec] (executor-thread-35) Skipping reading Query result cache data: cache-enabled = false, cache-mode = NORMAL
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Initializer list is empty
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.SQL] (executor-thread-35) select cscme1_0.SCOPE_ID from CLIENT_SCOPE_CLIENT cscme1_0 where cscme1_0.CLIENT_ID=? and cscme1_0.DEFAULT_SCOPE=?
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [6598b80c-7075-41b0-9146-7fbbb835939d]
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [6dd8def1-9012-468c-806a-04f208e66301]
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [ca5dd448-4f8d-4199-82c1-f8c185ad1263]
2024-10-10 13:00:19 2024-10-10 10:00:19,756 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,756 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [a125d956-9424-4dd5-af53-d81b5213402d]
2024-10-10 13:00:19 2024-10-10 10:00:19,760 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: client-jwt
2024-10-10 13:00:19 2024-10-10 10:00:19,773 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: client-secret-jwt
2024-10-10 13:00:19 2024-10-10 10:00:19,773 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: client-x509
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator CHALLENGE: client-x509
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: sending challenge for authentication execution client-x509
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) new JtaTransactionWrapper
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) was existing? true
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) JtaTransactionWrapper  commit
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) JtaTransactionWrapper end
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) JtaTransactionWrapper resuming suspended
2024-10-10 13:00:19 2024-10-10 10:00:19,779 WARN  [org.keycloak.events] (executor-thread-35) type="CLIENT_LOGIN_ERROR", realmId="0d94b14a-34ff-4f1d-b07a-dca1456772f4", clientId="qwerty", userId="null", ipAddress="192.168.65.1", error="invalid_client_credentials", grant_type="client_credentials"
2024-10-10 13:00:19 2024-10-10 10:00:19,779 DEBUG [WebApplicationException] (executor-thread-35) Restarting handler chain for exception exception: jakarta.ws.rs.WebApplicationException: HTTP 400 Bad Request
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.utils.AuthorizeClientUtil.authorizeClient(AuthorizeClientUtil.java:57)
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.checkClient(TokenEndpoint.java:168)
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:134)
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.endpoints.TokenEndpoint$quarkusrestinvoker$processGrantRequest_6408e15340992839b66447750c221d9aaa837bd7.invoke(Unknown Source)
2024-10-10 13:00:19     at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
2024-10-10 13:00:19     at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
2024-10-10 13:00:19     at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
2024-10-10 13:00:19     at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
2024-10-10 13:00:19     at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
2024-10-10 13:00:19     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
2024-10-10 13:00:19     at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
2024-10-10 13:00:19     at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
2024-10-10 13:00:19     at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2024-10-10 13:00:19     at java.base/java.lang.Thread.run(Thread.java:840)
2024-10-10 13:00:19 
```

On client:

```
Caused by: jakarta.ws.rs.ProcessingException: jakarta.ws.rs.BadRequestException: HTTP 400 Bad Request
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:652) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:424) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:134) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:61) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at jdk.proxy2/jdk.proxy2.$Proxy46.getInfo(Unknown Source) ~[na:na]
	at io.github.igpetrov.x509kc.X509kcApplication.keycloakAdmin(X509kcApplication.java:53) ~[main/:na]
	at io.github.igpetrov.x509kc.X509kcApplication$$SpringCGLIB$$0.CGLIB$keycloakAdmin$1(<generated>) ~[main/:na]
	at io.github.igpetrov.x509kc.X509kcApplication$$SpringCGLIB$$FastClass$$1.invoke(<generated>) ~[main/:na]
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258) ~[spring-core-6.1.13.jar:6.1.13]
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:348) ~[spring-context-6.1.13.jar:6.1.13]
	at io.github.igpetrov.x509kc.X509kcApplication$$SpringCGLIB$$0.keycloakAdmin(<generated>) ~[main/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:146) ~[spring-beans-6.1.13.jar:6.1.13]
	... 19 common frames omitted
Caused by: jakarta.ws.rs.BadRequestException: HTTP 400 Bad Request
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:236) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.extractResult(ClientInvocation.java:216) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.BodyEntityExtractor.extractEntity(BodyEntityExtractor.java:59) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:136) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:61) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at jdk.proxy2/jdk.proxy2.$Proxy44.grantToken(Unknown Source) ~[na:na]
	at org.keycloak.admin.client.token.TokenManager.grantToken(TokenManager.java:99) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.keycloak.admin.client.token.TokenManager.getAccessToken(TokenManager.java:75) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.keycloak.admin.client.token.TokenManager.getAccessTokenString(TokenManager.java:70) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.keycloak.admin.client.resource.BearerAuthFilter.filter(BearerAuthFilter.java:52) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:644) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	... 33 common frames omitted
```

At the same time, if I `curl` KC with the same certificates, everything seems to be alright, e.g.:

```
curl -X POST "https://localhost:18443/auth/realms/master/protocol/openid-connect/token" \
                                  --cert ./certs/personal-crt.pem \
                                  --key ./certs/personal-key.pem \
                                  --cacert ./certs/rootCA.pem \
                                  -d "client_id=qwerty" \
                                  -d "client_secret=xxxxxxx" \
                                  -d "grant_type=client_credentials" \
                                  -H "Content-Type: application/x-www-form-urlencoded" \
                                    -vvv --insecure
```

In terminal:

```
* Host localhost:18443 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:18443...
* Connected to localhost (::1) port 18443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Request CERT (13):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Certificate (11):
* (304) (OUT), TLS handshake, CERT verify (15):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: C=DE; ST=BW; L=Karlsruhe; O=IDP Security Provider GmbH; OU=Keycloak Dept; CN=keycloak
*  start date: Oct 10 08:24:06 2024 GMT
*  expire date: Oct 10 08:24:06 2025 GMT
*  issuer: C=CA; ST=BC; L=Vancouver; O=Trustworthy CA Provider LTD; OU=Trustworthy CA Provider LTD; CN=Root CA
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://localhost:18443/auth/realms/master/protocol/openid-connect/token
* [HTTP/2] [1] [:method: POST]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: localhost:18443]
* [HTTP/2] [1] [:path: /auth/realms/master/protocol/openid-connect/token]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
* [HTTP/2] [1] [content-type: application/x-www-form-urlencoded]
* [HTTP/2] [1] [content-length: 68]
> POST /auth/realms/master/protocol/openid-connect/token HTTP/2
> Host: localhost:18443
> User-Agent: curl/8.7.1
> Accept: */*
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 68
> 
* upload completely sent off: 68 bytes
< HTTP/2 200 
< cache-control: no-store
< pragma: no-cache
< content-length: 1400
< content-type: application/json
< referrer-policy: no-referrer
< strict-transport-security: max-age=31536000; includeSubDomains
< x-content-type-options: nosniff
< x-frame-options: SAMEORIGIN
< x-xss-protection: 1; mode=block
< 
* Connection #0 to host localhost left intact
{"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJwMm9hbEpGQXFqWXFsQWRQQzdqdVI4WVp3SVpKQ19NdzJOZXhEejJPb3NZIn0.eyJleHAiOjE3Mjg1NTQ3NTMsImlhdCI6MTcyODU1NDY5MywianRpIjoiY2QwNmViOWMtZjllZi00NzYzLTljYzctNDhmYjQ3YTc4YWU5IiwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6MTg0NDMvYXV0aC9yZWFsbXMvbWFzdGVyIiwiYXVkIjoiYWNjb3VudCIsInN1YiI6IjEzY2ViOGZmLTBjOTEtNDUxNS1iMTkyLTNkNjJhYTNlNjg2OSIsInR5cCI6IkJlYXJlciIsImF6cCI6InF3ZXJ0eSIsImFjciI6IjEiLCJhbGxvd2VkLW9yaWdpbnMiOlsiLyoiXSwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbImRlZmF1bHQtcm9sZXMtbWFzdGVyIiwib2ZmbGluZV9hY2Nlc3MiLCJ1bWFfYXV0aG9yaXphdGlvbiJdfSwicmVzb3VyY2VfYWNjZXNzIjp7ImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyIsInZpZXctcHJvZmlsZSJdfX0sInNjb3BlIjoiZW1haWwgcHJvZmlsZSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwiY2xpZW50SG9zdCI6IjE5Mi4xNjguNjUuMSIsInByZWZlcnJlZF91c2VybmFtZSI6InNlcnZpY2UtYWNjb3VudC1xd2VydHkiLCJjbGllbnRBZGRyZXNzIjoiMTkyLjE2OC42NS4xIiwiY2xpZW50X2lkIjoicXdlcnR5In0.fhQRVZ9yY1zDpqoIPinKouuPTX5Hh6bqT6cKIg0cIVoRbOgO1bELrqNEurk7bCcT_uuFGL3-TvRIE2l9mTIRVjekZn_Xwmu0ol2QghZqSPjtPYt8EhMeXkDhTKh_67ix5TtY5iyoTT0CXqyKSwYIzxXfMvkvY-PMjCR1OS-n-v5D3xaBVPWYqi7clnl1ohuCpZlEeq9O_7ZNHFECF5xOiNilAkuC5d0F_wNoz9kwNHKIrZuEk8iRxoewscjIXhC9vmS3XCyHh9F21WuGHsx3qVLAhWDmRCGEPmsBO89BdDW7mOl4zZyopLH-epG4qyn8QvL-qaj54X-aXpHzjn4j1A","expires_in":60,"refresh_expires_in":0,"token_type":"Bearer","not-before-policy":0,"scope":"email profile"}
```

In Keycloak:

```
2024-10-10 13:06:20 2024-10-10 10:06:20,832 DEBUG [jdk.event.security] (vert.x-eventloop-thread-1) X509Certificate: Alg:SHA256withRSA, Serial:f009d26a193c1731, Subject:CN=Private, OU=Private, O=Private, L=Odesa, ST=ODS, C=UA, EMAILADDRESS=igpetrov@github.io, Issuer:CN=Root CA, OU=Trustworthy CA Provider LTD, O=Trustworthy CA Provider LTD, L=Vancouver, ST=BC, C=CA, Key type:RSA, Length:4096, Cert Id:193173050, Valid from:10/10/24, 8:24 AM, Valid until:10/10/25, 8:24 AM
2024-10-10 13:06:20 2024-10-10 10:06:20,835 DEBUG [jdk.event.security] (vert.x-eventloop-thread-1) ValidationChain: 1314603832, 193173050
2024-10-10 13:06:20 2024-10-10 10:06:20,844 DEBUG [jdk.event.security] (vert.x-eventloop-thread-1)  TLSHandshake: null:-1, TLSv1.3, TLS_AES_256_GCM_SHA384, 193173050
2024-10-10 13:06:20 2024-10-10 10:06:20,844 DEBUG [io.netty.handler.ssl.SslHandler] (vert.x-eventloop-thread-1) [id: 0xe6b8f60e, L:/172.20.0.3:18443 - R:/192.168.65.1:21958] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
2024-10-10 13:06:20 2024-10-10 10:06:20,848 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) new JtaTransactionWrapper
2024-10-10 13:06:20 2024-10-10 10:06:20,848 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) was existing? false
2024-10-10 13:06:20 2024-10-10 10:06:20,853 DEBUG [org.keycloak.authentication.AuthenticationProcessor] (executor-thread-1) AUTHENTICATE CLIENT
2024-10-10 13:06:20 2024-10-10 10:06:20,853 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-secret
2024-10-10 13:06:20 2024-10-10 10:06:20,854 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-jwt
2024-10-10 13:06:20 2024-10-10 10:06:20,854 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-secret-jwt
2024-10-10 13:06:20 2024-10-10 10:06:20,854 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-x509
2024-10-10 13:06:20 2024-10-10 10:06:20,855 DEBUG [org.keycloak.services] (executor-thread-1) [X509ClientCertificateAuthenticator:authenticate] Matched CN=Private, OU=Private, O=Private, L=Odesa, ST=ODS, C=UA, EMAILADDRESS=igpetrov@github.io certificate.
2024-10-10 13:06:20 2024-10-10 10:06:20,855 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator SUCCESS: client-x509
2024-10-10 13:06:20 2024-10-10 10:06:20,855 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) Client qwerty authenticated by client-x509
2024-10-10 13:06:20 2024-10-10 10:06:20,861 DEBUG [org.keycloak.services.managers.AuthenticationSessionManager] (executor-thread-1) Removing root authSession '66f90c5f-e857-4fc1-87fe-22f890c8ed15'. Expire restart cookie: true
2024-10-10 13:06:20 2024-10-10 10:06:20,869 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) JtaTransactionWrapper  commit
2024-10-10 13:06:20 2024-10-10 10:06:20,869 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) JtaTransactionWrapper end
2024-10-10 13:06:20 2024-10-10 10:06:20,869 DEBUG [org.keycloak.events] (executor-thread-1) type="CLIENT_LOGIN", realmId="0d94b14a-34ff-4f1d-b07a-dca1456772f4", clientId="qwerty", userId="13ceb8ff-0c91-4515-b192-3d62aa3e6869", sessionId="66f90c5f-e857-4fc1-87fe-22f890c8ed15", ipAddress="192.168.65.1", token_id="8ed02328-1486-43cb-9002-d73840ee3af9", grant_type="client_credentials", scope="email profile", client_auth_method="client-x509", username="service-account-qwerty", authSessionParentId="66f90c5f-e857-4fc1-87fe-22f890c8ed15", authSessionTabId="JLicr5ifYcY"
```

### Version

24.0.5, 26.0.0

### Regression

- [ ] The issue is a regression

### Expected behavior

Keycloak client does not fail with HTTP status `400 Bad Request`, similar to `curl` (copied from description):

```
curl -X POST "https://localhost:18443/auth/realms/master/protocol/openid-connect/token" \
                                  --cert ./certs/personal-crt.pem \
                                  --key ./certs/personal-key.pem \
                                  --cacert ./certs/rootCA.pem \
                                  -d "client_id=qwerty" \
                                  -d "client_secret=xxxxxxx" \
                                  -d "grant_type=client_credentials" \
                                  -H "Content-Type: application/x-www-form-urlencoded" \
                                    -vvv --insecure
```

In terminal:

```
* Host localhost:18443 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:18443...
* Connected to localhost (::1) port 18443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Request CERT (13):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Certificate (11):
* (304) (OUT), TLS handshake, CERT verify (15):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: C=DE; ST=BW; L=Karlsruhe; O=IDP Security Provider GmbH; OU=Keycloak Dept; CN=keycloak
*  start date: Oct 10 08:24:06 2024 GMT
*  expire date: Oct 10 08:24:06 2025 GMT
*  issuer: C=CA; ST=BC; L=Vancouver; O=Trustworthy CA Provider LTD; OU=Trustworthy CA Provider LTD; CN=Root CA
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://localhost:18443/auth/realms/master/protocol/openid-connect/token
* [HTTP/2] [1] [:method: POST]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: localhost:18443]
* [HTTP/2] [1] [:path: /auth/realms/master/protocol/openid-connect/token]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
* [HTTP/2] [1] [content-type: application/x-www-form-urlencoded]
* [HTTP/2] [1] [content-length: 68]
> POST /auth/realms/master/protocol/openid-connect/token HTTP/2
> Host: localhost:18443
> User-Agent: curl/8.7.1
> Accept: */*
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 68
> 
* upload completely sent off: 68 bytes
< HTTP/2 200 
< cache-control: no-store
< pragma: no-cache
< content-length: 1400
< content-type: application/json
< referrer-policy: no-referrer
< strict-transport-security: max-age=31536000; includeSubDomains
< x-content-type-options: nosniff
< x-frame-options: SAMEORIGIN
< x-xss-protection: 1; mode=block
< 
* Connection #0 to host localhost left intact
{"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJwMm9hbEpGQXFqWXFsQWRQQzdqdVI4WVp3SVpKQ19NdzJOZXhEejJPb3NZIn0.eyJleHAiOjE3Mjg1NTQ3NTMsImlhdCI6MTcyODU1NDY5MywianRpIjoiY2QwNmViOWMtZjllZi00NzYzLTljYzctNDhmYjQ3YTc4YWU5IiwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6MTg0NDMvYXV0aC9yZWFsbXMvbWFzdGVyIiwiYXVkIjoiYWNjb3VudCIsInN1YiI6IjEzY2ViOGZmLTBjOTEtNDUxNS1iMTkyLTNkNjJhYTNlNjg2OSIsInR5cCI6IkJlYXJlciIsImF6cCI6InF3ZXJ0eSIsImFjciI6IjEiLCJhbGxvd2VkLW9yaWdpbnMiOlsiLyoiXSwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbImRlZmF1bHQtcm9sZXMtbWFzdGVyIiwib2ZmbGluZV9hY2Nlc3MiLCJ1bWFfYXV0aG9yaXphdGlvbiJdfSwicmVzb3VyY2VfYWNjZXNzIjp7ImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyIsInZpZXctcHJvZmlsZSJdfX0sInNjb3BlIjoiZW1haWwgcHJvZmlsZSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwiY2xpZW50SG9zdCI6IjE5Mi4xNjguNjUuMSIsInByZWZlcnJlZF91c2VybmFtZSI6InNlcnZpY2UtYWNjb3VudC1xd2VydHkiLCJjbGllbnRBZGRyZXNzIjoiMTkyLjE2OC42NS4xIiwiY2xpZW50X2lkIjoicXdlcnR5In0.fhQRVZ9yY1zDpqoIPinKouuPTX5Hh6bqT6cKIg0cIVoRbOgO1bELrqNEurk7bCcT_uuFGL3-TvRIE2l9mTIRVjekZn_Xwmu0ol2QghZqSPjtPYt8EhMeXkDhTKh_67ix5TtY5iyoTT0CXqyKSwYIzxXfMvkvY-PMjCR1OS-n-v5D3xaBVPWYqi7clnl1ohuCpZlEeq9O_7ZNHFECF5xOiNilAkuC5d0F_wNoz9kwNHKIrZuEk8iRxoewscjIXhC9vmS3XCyHh9F21WuGHsx3qVLAhWDmRCGEPmsBO89BdDW7mOl4zZyopLH-epG4qyn8QvL-qaj54X-aXpHzjn4j1A","expires_in":60,"refresh_expires_in":0,"token_type":"Bearer","not-before-policy":0,"scope":"email profile"}
```

In Keycloak:

```
2024-10-10 13:06:20 2024-10-10 10:06:20,832 DEBUG [jdk.event.security] (vert.x-eventloop-thread-1) X509Certificate: Alg:SHA256withRSA, Serial:f009d26a193c1731, Subject:CN=Private, OU=Private, O=Private, L=Odesa, ST=ODS, C=UA, EMAILADDRESS=igpetrov@github.io, Issuer:CN=Root CA, OU=Trustworthy CA Provider LTD, O=Trustworthy CA Provider LTD, L=Vancouver, ST=BC, C=CA, Key type:RSA, Length:4096, Cert Id:193173050, Valid from:10/10/24, 8:24 AM, Valid until:10/10/25, 8:24 AM
2024-10-10 13:06:20 2024-10-10 10:06:20,835 DEBUG [jdk.event.security] (vert.x-eventloop-thread-1) ValidationChain: 1314603832, 193173050
2024-10-10 13:06:20 2024-10-10 10:06:20,844 DEBUG [jdk.event.security] (vert.x-eventloop-thread-1)  TLSHandshake: null:-1, TLSv1.3, TLS_AES_256_GCM_SHA384, 193173050
2024-10-10 13:06:20 2024-10-10 10:06:20,844 DEBUG [io.netty.handler.ssl.SslHandler] (vert.x-eventloop-thread-1) [id: 0xe6b8f60e, L:/172.20.0.3:18443 - R:/192.168.65.1:21958] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_256_GCM_SHA384
2024-10-10 13:06:20 2024-10-10 10:06:20,848 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) new JtaTransactionWrapper
2024-10-10 13:06:20 2024-10-10 10:06:20,848 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) was existing? false
2024-10-10 13:06:20 2024-10-10 10:06:20,853 DEBUG [org.keycloak.authentication.AuthenticationProcessor] (executor-thread-1) AUTHENTICATE CLIENT
2024-10-10 13:06:20 2024-10-10 10:06:20,853 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-secret
2024-10-10 13:06:20 2024-10-10 10:06:20,854 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-jwt
2024-10-10 13:06:20 2024-10-10 10:06:20,854 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-secret-jwt
2024-10-10 13:06:20 2024-10-10 10:06:20,854 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator: client-x509
2024-10-10 13:06:20 2024-10-10 10:06:20,855 DEBUG [org.keycloak.services] (executor-thread-1) [X509ClientCertificateAuthenticator:authenticate] Matched CN=Private, OU=Private, O=Private, L=Odesa, ST=ODS, C=UA, EMAILADDRESS=igpetrov@github.io certificate.
2024-10-10 13:06:20 2024-10-10 10:06:20,855 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) client authenticator SUCCESS: client-x509
2024-10-10 13:06:20 2024-10-10 10:06:20,855 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-1) Client qwerty authenticated by client-x509
2024-10-10 13:06:20 2024-10-10 10:06:20,861 DEBUG [org.keycloak.services.managers.AuthenticationSessionManager] (executor-thread-1) Removing root authSession '66f90c5f-e857-4fc1-87fe-22f890c8ed15'. Expire restart cookie: true
2024-10-10 13:06:20 2024-10-10 10:06:20,869 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) JtaTransactionWrapper  commit
2024-10-10 13:06:20 2024-10-10 10:06:20,869 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-1) JtaTransactionWrapper end
2024-10-10 13:06:20 2024-10-10 10:06:20,869 DEBUG [org.keycloak.events] (executor-thread-1) type="CLIENT_LOGIN", realmId="0d94b14a-34ff-4f1d-b07a-dca1456772f4", clientId="qwerty", userId="13ceb8ff-0c91-4515-b192-3d62aa3e6869", sessionId="66f90c5f-e857-4fc1-87fe-22f890c8ed15", ipAddress="192.168.65.1", token_id="8ed02328-1486-43cb-9002-d73840ee3af9", grant_type="client_credentials", scope="email profile", client_auth_method="client-x509", username="service-account-qwerty", authSessionParentId="66f90c5f-e857-4fc1-87fe-22f890c8ed15", authSessionTabId="JLicr5ifYcY"
```

### Actual behavior

Client fails with HTTP status code 400 Bad Request (copied from description):

```
final Keycloak keycloakClient = KeycloakBuilder.builder()
        .serverUrl("https://localhost:18443/auth")
        .realm("master")
        .grantType(OAuth2Constants.CLIENT_CREDENTIALS)
        .clientId("qwerty")
        .clientSecret("xxxxxxx")
        .resteasyClient(resteasyClient)
        .build();
    
    // dummy call
    final var serverInfo = keycloakClient.serverInfo().getInfo();
```

On the Keycloak:

```
2024-10-10 13:00:19 2024-10-10 10:00:19,754 DEBUG [org.hibernate.internal.util.EntityPrinter] (executor-thread-35) org.keycloak.models.jpa.entities.ClientAttributeEntity{value=CN=Private, OU=Private, O=Private, L=Odesa, ST=ODS, C=UA, EMAILADDRESS=igpetrov@github.io}
2024-10-10 13:00:19 2024-10-10 10:00:19,754 DEBUG [org.hibernate.orm.sql.exec] (executor-thread-35) Skipping reading Query result cache data: cache-enabled = false, cache-mode = NORMAL
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Initializer list is empty
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.SQL] (executor-thread-35) select cscme1_0.SCOPE_ID from CLIENT_SCOPE_CLIENT cscme1_0 where cscme1_0.CLIENT_ID=? and cscme1_0.DEFAULT_SCOPE=?
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [6598b80c-7075-41b0-9146-7fbbb835939d]
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [6dd8def1-9012-468c-806a-04f208e66301]
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,755 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [ca5dd448-4f8d-4199-82c1-f8c185ad1263]
2024-10-10 13:00:19 2024-10-10 10:00:19,756 DEBUG [org.hibernate.orm.results.loading] (executor-thread-35) Calling top-level assembler (0 / 1) : org.hibernate.sql.results.graph.basic.BasicResultAssembler@1d88414a
2024-10-10 13:00:19 2024-10-10 10:00:19,756 DEBUG [org.hibernate.orm.results] (executor-thread-35) Extracted JDBC value [0] - [a125d956-9424-4dd5-af53-d81b5213402d]
2024-10-10 13:00:19 2024-10-10 10:00:19,760 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: client-jwt
2024-10-10 13:00:19 2024-10-10 10:00:19,773 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: client-secret-jwt
2024-10-10 13:00:19 2024-10-10 10:00:19,773 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: client-x509
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator CHALLENGE: client-x509
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.authentication.ClientAuthenticationFlow] (executor-thread-35) client authenticator: sending challenge for authentication execution client-x509
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) new JtaTransactionWrapper
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) was existing? true
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) JtaTransactionWrapper  commit
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) JtaTransactionWrapper end
2024-10-10 13:00:19 2024-10-10 10:00:19,778 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-35) JtaTransactionWrapper resuming suspended
2024-10-10 13:00:19 2024-10-10 10:00:19,779 WARN  [org.keycloak.events] (executor-thread-35) type="CLIENT_LOGIN_ERROR", realmId="0d94b14a-34ff-4f1d-b07a-dca1456772f4", clientId="qwerty", userId="null", ipAddress="192.168.65.1", error="invalid_client_credentials", grant_type="client_credentials"
2024-10-10 13:00:19 2024-10-10 10:00:19,779 DEBUG [WebApplicationException] (executor-thread-35) Restarting handler chain for exception exception: jakarta.ws.rs.WebApplicationException: HTTP 400 Bad Request
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.utils.AuthorizeClientUtil.authorizeClient(AuthorizeClientUtil.java:57)
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.checkClient(TokenEndpoint.java:168)
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:134)
2024-10-10 13:00:19     at org.keycloak.protocol.oidc.endpoints.TokenEndpoint$quarkusrestinvoker$processGrantRequest_6408e15340992839b66447750c221d9aaa837bd7.invoke(Unknown Source)
2024-10-10 13:00:19     at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
2024-10-10 13:00:19     at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
2024-10-10 13:00:19     at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
2024-10-10 13:00:19     at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
2024-10-10 13:00:19     at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
2024-10-10 13:00:19     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
2024-10-10 13:00:19     at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
2024-10-10 13:00:19     at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
2024-10-10 13:00:19     at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
2024-10-10 13:00:19     at java.base/java.lang.Thread.run(Thread.java:840)
2024-10-10 13:00:19 
```

On client:

```
Caused by: jakarta.ws.rs.ProcessingException: jakarta.ws.rs.BadRequestException: HTTP 400 Bad Request
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:652) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:424) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:134) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:61) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at jdk.proxy2/jdk.proxy2.$Proxy46.getInfo(Unknown Source) ~[na:na]
	at io.github.igpetrov.x509kc.X509kcApplication.keycloakAdmin(X509kcApplication.java:53) ~[main/:na]
	at io.github.igpetrov.x509kc.X509kcApplication$$SpringCGLIB$$0.CGLIB$keycloakAdmin$1(<generated>) ~[main/:na]
	at io.github.igpetrov.x509kc.X509kcApplication$$SpringCGLIB$$FastClass$$1.invoke(<generated>) ~[main/:na]
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258) ~[spring-core-6.1.13.jar:6.1.13]
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:348) ~[spring-context-6.1.13.jar:6.1.13]
	at io.github.igpetrov.x509kc.X509kcApplication$$SpringCGLIB$$0.keycloakAdmin(<generated>) ~[main/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:146) ~[spring-beans-6.1.13.jar:6.1.13]
	... 19 common frames omitted
Caused by: jakarta.ws.rs.BadRequestException: HTTP 400 Bad Request
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:236) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.extractResult(ClientInvocation.java:216) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.BodyEntityExtractor.extractEntity(BodyEntityExtractor.java:59) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:136) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:61) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	at jdk.proxy2/jdk.proxy2.$Proxy44.grantToken(Unknown Source) ~[na:na]
	at org.keycloak.admin.client.token.TokenManager.grantToken(TokenManager.java:99) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.keycloak.admin.client.token.TokenManager.getAccessToken(TokenManager.java:75) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.keycloak.admin.client.token.TokenManager.getAccessTokenString(TokenManager.java:70) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.keycloak.admin.client.resource.BearerAuthFilter.filter(BearerAuthFilter.java:52) ~[keycloak-admin-client-24.0.5.jar:24.0.5]
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:644) ~[resteasy-client-6.2.7.Final.jar:6.2.7.Final]
	... 33 common frames omitted
```

### How to Reproduce?

## Conceptual

1. Create root, keycloak, and client X509 certificates
1. Start a clean KC instance, e.g. v.26.0.0 in Docker container
1. Login into KC as admin user
1. Create a new client with any `Client ID`
1. In the `Capability config` section, tick `Client authentication` and `Service account roles`
1. Once the client is created, go to `Credentials` tab, and select `X509 Certificate` in the `Client Authenticator` field
1. Input correct `Subject DN` of the client, based on the generated (and signed!) certificate
1. `curl` the token endpoint with the certificates to verify everything works with `curl`
1. Create a new Keycloak class instance in Java, e.g. as `keycloakClient = KeycloakBuilder.builder().serverUrl("...").realm("...").grantType(CLIENT_CREDENTIALS).clientId("...").clientSecret("...xx").resteasyClient(resteasyClient).build();`
1. Make any arbitrary call, e.g. `keycloakClient.serverInfo().getInfo()`.
1. Observe HTTP 400 bad request in both Keycloak and Java client.

## Concrete example

Please refer [this repository](https://github.com/igpetrov/kc-x509). It is a minimal code version to reproduce the issue.

### Clone repository

The repository https://github.com/igpetrov/kc-x509 contains script to generate X509 certificates, docker compose file, and ready-to-use minimal application code that reproduces the issue.

### Generate certificate

Generate new with the [script](https://github.com/igpetrov/kc-x509/blob/main/certs/generate.sh) or use [existing ones](https://github.com/igpetrov/kc-x509/tree/main/certs).

### Start Keycloak

Start [Keycloak and Postgres](https://github.com/igpetrov/kc-x509/blob/main/docker-compose.yaml) with `docker-compose up`

### Create a new KC client

I did it like this.

![Screenshot 2024-10-10 at 13 44 39](https://github.com/user-attachments/assets/7ac4582e-49cd-48c1-9cb0-1a8f28f849f0)
![Screenshot 2024-10-10 at 13 44 53](https://github.com/user-attachments/assets/6e135c57-6cac-499b-920f-bb41dd03a819)
![Screenshot 2024-10-10 at 13 45 01](https://github.com/user-attachments/assets/51798b66-3357-46bd-848c-feff70692a65)
![Screenshot 2024-10-10 at 13 45 34](https://github.com/user-attachments/assets/dc550f35-fd57-401b-9622-f2da5ad566f7)


Note: DN `CN=Private, OU=Private, O=Private, L=Odesa, ST=ODS, C=UA, EMAILADDRESS=igpetrov@github.io` is the same from `personal` certificate.

### (optional) cURL to make sure token is created. This will give confidence that everything configured correctly.

```
curl -X POST "https://localhost:18443/auth/realms/master/protocol/openid-connect/token" \
  --cert ./certs/personal-crt.pem \
  --key ./certs/personal-key.pem \
  --cacert ./certs/rootCA.pem \
  -d "client_id=qwerty" \
  -d "client_secret=xxxxxxx" \
  -d "grant_type=client_credentials" \
  -H "Content-Type: application/x-www-form-urlencoded" \
    -vvv --insecure
Note: Unnecessary use of -X or --request, POST is already inferred.
* Host localhost:18443 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:18443...
* Connected to localhost (::1) port 18443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Request CERT (13):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Certificate (11):
* (304) (OUT), TLS handshake, CERT verify (15):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384 / [blank] / UNDEF
* ALPN: server accepted h2
* Server certificate:
*  subject: C=DE; ST=BW; L=Karlsruhe; O=IDP Security Provider GmbH; OU=Keycloak Dept; CN=keycloak
*  start date: Oct 10 08:24:06 2024 GMT
*  expire date: Oct 10 08:24:06 2025 GMT
*  issuer: C=CA; ST=BC; L=Vancouver; O=Trustworthy CA Provider LTD; OU=Trustworthy CA Provider LTD; CN=Root CA
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://localhost:18443/auth/realms/master/protocol/openid-connect/token
* [HTTP/2] [1] [:method: POST]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: localhost:18443]
* [HTTP/2] [1] [:path: /auth/realms/master/protocol/openid-connect/token]
* [HTTP/2] [1] [user-agent: curl/8.7.1]
* [HTTP/2] [1] [accept: */*]
* [HTTP/2] [1] [content-type: application/x-www-form-urlencoded]
* [HTTP/2] [1] [content-length: 68]
> POST /auth/realms/master/protocol/openid-connect/token HTTP/2
> Host: localhost:18443
> User-Agent: curl/8.7.1
> Accept: */*
> Content-Type: application/x-www-form-urlencoded
> Content-Length: 68
> 
* upload completely sent off: 68 bytes
< HTTP/2 200 
< cache-control: no-store
< pragma: no-cache
< content-length: 1400
< content-type: application/json
< referrer-policy: no-referrer
< strict-transport-security: max-age=31536000; includeSubDomains
< x-content-type-options: nosniff
< x-frame-options: SAMEORIGIN
< x-xss-protection: 1; mode=block
< 
* Connection #0 to host localhost left intact
{"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI3aXZGVGlXeGhWYlZiTzJkRFNQUTZCanhiSXlSN2haVEhpSWpFMllIR3l3In0.eyJleHAiOjE3Mjg1NTc0NTksImlhdCI6MTcyODU1NzM5OSwianRpIjoiNzY0YzVmZTgtYWUyYi00MDlmLThjNTktY2I1NTllYTFiMzVjIiwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6MTg0NDMvYXV0aC9yZWFsbXMvbWFzdGVyIiwiYXVkIjoiYWNjb3VudCIsInN1YiI6ImFhZTY5NTAxLWM3OTEtNDY5My1iYjBiLTRhYWI1N2RlMjA5ZSIsInR5cCI6IkJlYXJlciIsImF6cCI6InF3ZXJ0eSIsImFjciI6IjEiLCJhbGxvd2VkLW9yaWdpbnMiOlsiLyoiXSwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbImRlZmF1bHQtcm9sZXMtbWFzdGVyIiwib2ZmbGluZV9hY2Nlc3MiLCJ1bWFfYXV0aG9yaXphdGlvbiJdfSwicmVzb3VyY2VfYWNjZXNzIjp7ImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyIsInZpZXctcHJvZmlsZSJdfX0sInNjb3BlIjoiZW1haWwgcHJvZmlsZSIsImNsaWVudEhvc3QiOiIxOTIuMTY4LjY1LjEiLCJlbWFpbF92ZXJpZmllZCI6ZmFsc2UsInByZWZlcnJlZF91c2VybmFtZSI6InNlcnZpY2UtYWNjb3VudC1xd2VydHkiLCJjbGllbnRBZGRyZXNzIjoiMTkyLjE2OC42NS4xIiwiY2xpZW50X2lkIjoicXdlcnR5In0.L2OAUDbYO_McL6oVW5fB7PCZIcwDRavm-v2eSA3hTmxeO2P94CvpfYmLTfxrx81qP9YCeM6cbwXI_GhxAArcHz9SGJlXlLMaTDGg5PX-glAX1lcpehMSSq26of7L1sszanwHy4uJC6Pt0zLMcRQyqYF8py8Xgy_AqAkhqAtZTCYNflqZQGVX1DsZCvZJpW_eGwxyR3E658YCmJDnzJ2azx--fyimyJXHcjAeDxJ3QRp9amAUprRQo-f2Lwt_doW8DNXW0jz1cmkmpGhdFMeOY9g3p2tzGrDVStLBqiqGyxCCAyZY80KwETh5xGbukUd_MaFlnZ-zEYb0qTZVC2PymQ","expires_in":60,"refresh_expires_in":0,"token_type":"Bearer","not-before-policy":0,"scope":"email profile"}
```

Note: the certificates from example are the [same as mentioned earlier](https://github.com/igpetrov/kc-x509/tree/main/certs).

### Start Spring Boot Application

To reproduce, start the [Spring Boot application](https://github.com/igpetrov/kc-x509/blob/main/x509kc/src/main/java/io/github/igpetrov/x509kc/X509kcApplication.java). To path certificates and print additional debug information, I pass the following arguments:

```
-Djavax.net.ssl.trustStore=<PATH_TO_PROJECT>/certs/truststore.jks
-Djavax.net.ssl.trustStorePassword=password
-Djavax.net.ssl.keyStore=<PATH_TO_PROJECT>/certs/personal.p12
-Djavax.net.ssl.keyStorePassword=password
-Djdk.internal.httpclient.disableHostnameVerification=true
-Dcom.sun.jndi.ldap.object.disableEndpointIdentification=true
-Djavax.net.debug=ssl:handshake:verbose:keymanager:sslctx
```

Note: in IntelliJ you can do it by setting VM options:

![Screenshot 2024-10-10 at 13 58 26](https://github.com/user-attachments/assets/488c5877-6558-48b8-b069-d91d2ee4d4a7)


It will fail right away with HTTP 400 Bad Request. Then observe both logs in both Spring Boot application and Keycloak.

### Anything else?

Maybe some naїve observation, but this is what I found.

At some point, I observe the following call, where token is requested only with `grant_type`:

![Screenshot 2024-10-10 at 14 04 05](https://github.com/user-attachments/assets/c68238b7-4e7e-4399-b62a-e87f3a9473af)

And as you can see, the form does not contain the Client ID. Then, I found out that the client ID is only inserted when the [client is public](https://github.com/keycloak/keycloak/blob/release/26.0/integration/admin-client/src/main/java/org/keycloak/admin/client/token/TokenManager.java#L93), while "public" means there is [no client secret](https://github.com/keycloak/keycloak/blob/release/26.0/integration/admin-client/src/main/java/org/keycloak/admin/client/Config.java#L101), which totally makes sense, because I don't need secret anymore, since I am using X509.

Which means, that creating a client in a way, may not be correct:

```
final Keycloak keycloakClient = KeycloakBuilder.builder()
        .serverUrl("https://localhost:18443/auth")
        .realm("master")
        .grantType(OAuth2Constants.CLIENT_CREDENTIALS)
        .clientId("qwerty")
        .clientSecret("xxxxxxx") // <----- REDUNDANT?
        .resteasyClient(resteasyClient)
        .build();
```

But I cannot exclude `clientSecret(...)` builder call, because I am [not allowed](https://github.com/keycloak/keycloak/blob/release/26.0/integration/admin-client/src/main/java/org/keycloak/admin/client/KeycloakBuilder.java#L142-L146).

In my opinion, you should allow not setting `clientSecret` in the builder by eliminating [this statement](https://github.com/keycloak/keycloak/blob/release/26.0/integration/admin-client/src/main/java/org/keycloak/admin/client/KeycloakBuilder.java#L142-L146).

Potential fix: https://github.com/keycloak/keycloak/pull/33796

--- a/integration/admin-client/src/main/java/org/keycloak/admin/client/KeycloakBuilder.java
+++ b/integration/admin-client/src/main/java/org/keycloak/admin/client/KeycloakBuilder.java
@@ -139,10 +139,6 @@ public class KeycloakBuilder {
             if (password == null) {
                 throw new IllegalStateException("password required");
             }
-        } else if (CLIENT_CREDENTIALS.equals(grantType)) {
-            if (clientSecret == null) {
-                throw new IllegalStateException("clientSecret required with grant_type=client_credentials");
-            }
         }
 
         if (authorization == null && clientId == null) {

--- a/integration/admin-client/src/main/java/org/keycloak/admin/client/token/TokenManager.java
+++ b/integration/admin-client/src/main/java/org/keycloak/admin/client/token/TokenManager.java
@@ -60,10 +60,6 @@ public class TokenManager {
         }
         this.tokenService = Keycloak.getClientProvider().targetProxy(target, TokenService.class);
         this.accessTokenGrantType = config.getGrantType();
-
-        if (CLIENT_CREDENTIALS.equals(accessTokenGrantType) && config.isPublicClient()) {
-            throw new IllegalArgumentException("Can't use " + GRANT_TYPE + "=" + CLIENT_CREDENTIALS + " with public client");
-        }
     }
 
     public String getAccessTokenString() {

--- a/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/AdminClientUtil.java
+++ b/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/AdminClientUtil.java
@@ -18,10 +18,13 @@
 package org.keycloak.testsuite.util;
 
 import java.io.File;
+import java.io.FileInputStream;
 import java.io.IOException;
 import java.security.KeyManagementException;
+import java.security.KeyStore;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
 import java.security.cert.CertificateException;
 
 import javax.net.ssl.SSLContext;
@@ -99,6 +102,22 @@ public class AdminClientUtil {
                 .scope(scope).build();
     }
 
+    public static Keycloak createMTlsAdminClientWithClientCredentialsWithoutSecret(
+        final String realmName, String clientId, String scope)
+        throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException, KeyManagementException {
+
+        boolean ignoreUnknownProperties = false;
+        ResteasyClient resteasyClient = createResteasyClientWithKeystoreAndTruststore();
+
+        return KeycloakBuilder.builder()
+                .serverUrl(getAuthServerContextRoot() + "/auth")
+                .realm(realmName)
+                .grantType(OAuth2Constants.CLIENT_CREDENTIALS)
+                .clientId(clientId)
+                .resteasyClient(resteasyClient)
+                .scope(scope).build();
+  }
+
     public static Keycloak createAdminClient() throws Exception {
         return createAdminClient(false, getAuthServerContextRoot());
     }
@@ -145,6 +164,30 @@ public class AdminClientUtil {
         return resteasyClientBuilder.build();
     }
 
+    public static ResteasyClient createResteasyClientWithKeystoreAndTruststore() throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException, KeyManagementException {
+        ResteasyClientBuilder resteasyClientBuilder = (ResteasyClientBuilder) ResteasyClientBuilder.newBuilder();
+
+        if ("true".equals(System.getProperty("auth.server.ssl.required"))) {
+            File truststore = new File(PROJECT_BUILD_DIRECTORY, "dependency/keystore/keycloak.truststore");
+            try {
+                resteasyClientBuilder.sslContext(getSSLContextWithTruststoreAndKeystore(
+                  truststore, "secret",
+                  new File(PROJECT_BUILD_DIRECTORY, "dependency/keystore/keycloak.jks"), "secret"));
+            } catch (UnrecoverableKeyException e) {
+                throw new RuntimeException(e);
+            }
+
+            System.setProperty("javax.net.ssl.trustStore", truststore.getAbsolutePath());
+        }
+
+        resteasyClientBuilder
+            .hostnameVerification(ResteasyClientBuilder.HostnameVerificationPolicy.WILDCARD)
+            .connectionPoolSize(NUMBER_OF_CONNECTIONS)
+            .httpEngine(getCustomClientHttpEngine(resteasyClientBuilder, 1, null));
+
+        return resteasyClientBuilder.build();
+    }
+
     private static SSLContext getSSLContextWithTruststore(File file, String password) throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException, KeyManagementException {
         if (!file.isFile()) {
             throw new RuntimeException("Truststore file not found: " + file.getAbsolutePath());
@@ -153,6 +196,27 @@ public class AdminClientUtil {
                 .setProtocol("TLS")
                 .loadTrustMaterial(file, password == null ? null : password.toCharArray())
                 .build();
+        return theContext;
+    }
+
+    private static SSLContext getSSLContextWithTruststoreAndKeystore(
+        File trustStore, String truststorePassword, File keystore, String keystorePassword)
+        throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException, KeyManagementException, UnrecoverableKeyException {
+        if (!trustStore.isFile()) {
+            throw new RuntimeException("Truststore file not found: " + trustStore.getAbsolutePath());
+        }
+        if (!keystore.isFile()) {
+            throw new RuntimeException("Keystore file not found: " + keystore.getAbsolutePath());
+        }
+
+        KeyStore ks = KeyStore.getInstance("JKS");
+        ks.load(new FileInputStream(keystore), keystorePassword.toCharArray());
+
+        SSLContext theContext = SSLContexts.custom()
+            .setProtocol("TLS")
+            .loadTrustMaterial(trustStore, truststorePassword == null ? null : truststorePassword.toCharArray())
+            .loadKeyMaterial(ks, keystorePassword.toCharArray())
+            .build();
         return theContext;
     }
 

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/AdminClientTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/AdminClientTest.java
@@ -21,12 +21,14 @@ package org.keycloak.testsuite.admin;
 import java.util.List;
 import jakarta.ws.rs.NotAuthorizedException;
 import jakarta.ws.rs.core.Response;
+import java.util.Map;
 import org.junit.Assert;
 import org.junit.Rule;
 import org.junit.Test;
 import org.keycloak.admin.client.Keycloak;
 import org.keycloak.admin.client.resource.ClientResource;
 import org.keycloak.admin.client.resource.RealmResource;
+import org.keycloak.authentication.authenticators.client.X509ClientAuthenticator;
 import org.keycloak.common.constants.ServiceAccountConstants;
 import org.keycloak.models.AdminRoles;
 import org.keycloak.models.Constants;
@@ -52,13 +54,18 @@ import java.util.Objects;
 public class AdminClientTest extends AbstractKeycloakTest {
 
     private static String realmName;
-    
+
     private static String userId;
     private static String userName;
 
     private static String clientUUID;
     private static String clientId;
     private static String clientSecret;
+
+    private static String x509ClientUUID;
+    private static String x509ClientId;
+
+    private static String x509UserName;
 
     @Rule
     public AssertEvents events = new AssertEvents(this);
@@ -83,6 +90,17 @@ public class AdminClientTest extends AbstractKeycloakTest {
                 .build();
         realm.client(enabledAppWithSkipRefreshToken);
 
+        x509ClientId = "x509-client-sa";
+        ClientRepresentation x509ServiceAccountClient = ClientBuilder.create()
+              .clientId(x509ClientId)
+              .serviceAccountsEnabled(true)
+              .build();
+        x509ServiceAccountClient.setClientAuthenticatorType(X509ClientAuthenticator.PROVIDER_ID);
+        x509ServiceAccountClient.setAttributes(Map.of(
+            X509ClientAuthenticator.ATTR_SUBJECT_DN, "(.*?)(?:$)",
+            X509ClientAuthenticator.ATTR_ALLOW_REGEX_PATTERN_COMPARISON, "true"));
+        realm.client(x509ServiceAccountClient);
+
         userId = KeycloakModelUtils.generateId();
         userName = ServiceAccountConstants.SERVICE_ACCOUNT_USER_PREFIX + enabledAppWithSkipRefreshToken.getClientId();
         UserBuilder serviceAccountUser = UserBuilder.create()
@@ -90,6 +108,17 @@ public class AdminClientTest extends AbstractKeycloakTest {
                 .serviceAccountId(enabledAppWithSkipRefreshToken.getClientId())
                 .role(Constants.REALM_MANAGEMENT_CLIENT_ID, AdminRoles.REALM_ADMIN);
         realm.user(serviceAccountUser);
+
+        // This user is associated with the x509-client-sa service account above and
+        // give the service account a service account role "realm-management:realm-admin".
+        // Without the "realm-management:realm-admin" role we won't be able to test any actual
+        // admin call.
+        x509UserName = ServiceAccountConstants.SERVICE_ACCOUNT_USER_PREFIX + x509ServiceAccountClient.getClientId();
+        UserBuilder x509ServiceAccountUser = UserBuilder.create()
+            .username(x509UserName)
+            .serviceAccountId(x509ServiceAccountClient.getClientId())
+            .role(Constants.REALM_MANAGEMENT_CLIENT_ID, AdminRoles.REALM_ADMIN);
+        realm.user(x509ServiceAccountUser);
 
         UserBuilder defaultUser = UserBuilder.create()
                 .id(KeycloakModelUtils.generateId())
@@ -105,6 +134,7 @@ public class AdminClientTest extends AbstractKeycloakTest {
     public void importRealm(RealmRepresentation realm) {
         super.importRealm(realm);
         if (Objects.equals(realm.getRealm(), realmName)) {
+            x509ClientUUID = adminClient.realm(realmName).clients().findByClientId(x509ClientId).get(0).getId();
             clientUUID = adminClient.realm(realmName).clients().findByClientId(clientId).get(0).getId();
             userId = adminClient.realm(realmName).users().searchByUsername(userName, true).get(0).getId();
         }
@@ -183,12 +213,40 @@ public class AdminClientTest extends AbstractKeycloakTest {
             clientId, clientSecret, scopeName)) {
             final AccessTokenResponse accessToken = adminClient.tokenManager().getAccessToken();
             Assert.assertTrue(accessToken.getScope().contains(scopeName));
+            Assert.assertNotNull(adminClient.realm(realmName).clientScopes().get(scopeId).toRepresentation());
         }
         // without scope
         try (Keycloak adminClient = AdminClientUtil.createAdminClientWithClientCredentials(realmName,
             clientId, clientSecret, null)) {
             final AccessTokenResponse accessToken = adminClient.tokenManager().getAccessToken();
             Assert.assertFalse(accessToken.getScope().contains(scopeName));
+            Assert.assertNotNull(adminClient.realm(realmName).clientScopes().get(scopeId).toRepresentation());
+        }
+    }
+
+    // A client secret in not necessary when authentication is
+    // performed via X.509 authorizer.
+    @Test
+    public void noClientSecretWithClientCredentialsAuthSuccess() throws Exception {
+        final RealmResource testRealm = adminClient.realm(realmName);
+
+        final String scopeName = "dummyScope";
+        String scopeId = createScope(testRealm, scopeName, KeycloakModelUtils.generateId());
+        testRealm.clients().get(x509ClientUUID).addOptionalClientScope(scopeId);
+
+        // with scope and no client secret
+        try (Keycloak adminClient = AdminClientUtil.
+            createMTlsAdminClientWithClientCredentialsWithoutSecret(realmName, x509ClientId, scopeName)) {
+            final AccessTokenResponse accessToken = adminClient.tokenManager().getAccessToken();
+            Assert.assertTrue(accessToken.getScope().contains(scopeName));
+            Assert.assertNotNull(adminClient.realm(realmName).clientScopes().get(scopeId).toRepresentation());
+        }
+        // without scope and no client secret
+        try (Keycloak adminClient = AdminClientUtil.
+            createMTlsAdminClientWithClientCredentialsWithoutSecret(realmName, x509ClientId, null)) {
+            final AccessTokenResponse accessToken = adminClient.tokenManager().getAccessToken();
+            Assert.assertFalse(accessToken.getScope().contains(scopeName));
+            Assert.assertNotNull(adminClient.realm(realmName).clientScopes().get(scopeId).toRepresentation());
         }
     }
 

